sonic-net/sonic-buildimage
# [Mellanox] run module initialization when any SFP related API is called

#### Why I did it

Currently, there are a few issues related to module host management:

1. module initialization flow is triggered when `chassis.get_change_event` is called for the first time. It is too late which delays fast-reboot/warm-reboot convergence time
2. module initialization flow always send change event for all SFP objects to xcvrd even if there is no real module change event. It causes ports flapping during warm-reboot
3. legacy mode and module host management mode are mixed into the same code logic which makes it hard to maintain and extend

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it

To address above issues, the PR introduces following changes:

1. module initialization flow is triggered when chassis SFP related API is called for the first time. It means that any of the following API will trigger module init: `chassis.get_sfp`, `chassis.get_all_sfps`, `chassis.get_change_event`. The init flow shall only execute once under xcvrd context. 
2. module initialization flow does not send change event anymore. Change event shall be sent only if a real cable event is detected by `chassis.get_change_event` (plug-in/plug-out/error)
3. legacy mode and module host management mode are decoupled to avoid affecting each other

#### How to verify it

1. Unit test covered most of the new change
2. Full sonic-mgmt regression test make sure there is no degradation (test branch 202311)
3. Manual test 

#### Which release branch to backport (provide reason below if selected)




#### Tested branch (Please provide the tested image version)



- [x] 202311 2472123c6587ac26880d50411669e1bbfe89ea34

#### Description for the changelog




#### Link to config_db schema for YANG module changes


#### A picture of a cute animal (not mandatory but encouraged)
